RecoMuon/GlobalTrackingTools/src/MuonTrackingRegionBuilder.cc

```diff
@@ -1,281 +1,278 @@
+  theDeltaR = par.getParameter<double>("DeltaR");
+
+  // The static region size in Eta
+  theDeltaEta = par.getParameter<double>("DeltaEta");
+
+  // The static region size in Phi
+  theDeltaPhi = par.getParameter<double>("DeltaPhi");
+
+  // Maximum number of regions to build when looping over Muons
+  theMaxRegions = par.getParameter<int>("maxRegions");
+
+  // Flag to use precise??
+  thePrecise = par.getParameter<bool>("precise"); 
+
+  // perigee reference point ToDo: Check this
+  theOnDemand = RectangularEtaPhiTrackingRegion::intToUseMeasurementTracker(par.getParameter<int>("OnDemand"));
+  if (theOnDemand != RectangularEtaPhiTrackingRegion::UseMeasurementTracker::kNever) {
+    theMeasurementTrackerToken = iC.consumes<MeasurementTrackerEvent>(par.getParameter<edm::InputTag>("MeasurementTrackerName"));
+  }
+
+  // Vertex collection and Beam Spot
+  beamSpotToken = iC.consumes<reco::BeamSpot>(par.getParameter<edm::InputTag>("beamSpot"));
+  vertexCollectionToken = iC.consumes<reco::VertexCollection>(par.getParameter<edm::InputTag>("vertexCollection"));
+
+  // Input muon collection
+  inputCollectionToken = iC.consumes<reco::TrackCollection>(par.getParameter<edm::InputTag>("input"));
+}
+
+
+//
+// Member function to be compatible with TrackingRegionProducerFactory: create many ROI for many tracks
+//
+std::vector<TrackingRegion*> MuonTrackingRegionBuilder::regions(const edm::Event& ev, const edm::EventSetup& es) const {
+	
+  std::vector<TrackingRegion*> result;
+
+  edm::Handle<reco::TrackCollection> tracks;
+  ev.getByToken(inputCollectionToken, tracks);
+
+  int nRegions = 0;
+  for (auto it = tracks->cbegin(), ed = tracks->cend(); it != ed && nRegions < theMaxRegions; ++it) {
+    result.push_back(region(*it,ev));
+    nRegions++; 
+  }
+
+  return result;
+
+}
+
+
+//
+// Call region on Track from TrackRef
+//
+RectangularEtaPhiTrackingRegion* MuonTrackingRegionBuilder::region(const reco::TrackRef& track) const {
+  return region(*track);
+}
+
+
+//
+// ToDo: Not sure if this is needed?
+//
+void MuonTrackingRegionBuilder::setEvent(const edm::Event& event) {
+  theEvent = &event;
+}
+
+
+//
+//	Main member function called to create the ROI
+//
+RectangularEtaPhiTrackingRegion* MuonTrackingRegionBuilder::region(const reco::Track& staTrack, const edm::Event& ev) const {
+
+  // get track momentum/direction at vertex
+  const math::XYZVector& mom = staTrack.momentum();
+  GlobalVector dirVector(mom.x(),mom.y(),mom.z());
+  double pt = staTrack.pt();
+
+  // initial vertex position - in the following it is replaced with beamspot/vertexing
+  GlobalPoint vertexPos(0.0,0.0,0.0);
+  // standard 15.9, if useVertex than use error from  vertex
+  double deltaZ = theHalfZ;
+
+  // retrieve beam spot information
+  edm::Handle<reco::BeamSpot> bs;
+  bool bsHandleFlag = ev.getByToken(beamSpotToken, bs);
+
+  // check the validity, otherwise vertexing
+  if ( bsHandleFlag && bs.isValid() && !useVertex ) {

```

Comment:
If no beamSpot is found, is an empty collection put in the event,
or no collection at all? If the latter then the check and graceful 
handling of the product not existing in some events must be maintained!